lib/routes/token.js

```diff
@@ -175,9 +179,31 @@ module.exports = {
             });
         }
       } else if (params.grant_type === GRANT_REFRESH_TOKEN) {
-        return confirmClient(params.client_id, params.client_secret)
-        .then(function() {
-          return confirmRefreshToken(params);
+        var clientId = params.client_id;
+        // If the client has a client_secret, check that it's provided and valid in the refresh request
+        // If the client does not have client_secret, check that one was not provided in the refresh request
+        return db.getClient(buf(clientId)).then(function(dbClient) {
+          if (! dbClient) {
+            logger.debug('client.notFound', { id: clientId });
+            throw AppError.unknownClient(clientId);
+          }
+
+          var confirmClientPromise;
+
+          if (dbClient.publicClient) {
+            if (params.client_secret) {
+              throw new AppError.invalidRequestParameter('client_secret');
+            }
+
+            confirmClientPromise = confirmPkceClient(params.client_id);
+          } else {
+            confirmClientPromise = confirmClient(params.client_id, params.client_secret);
```

Comment:
The `confirmClient` does another db lookup, but this function already did one, and then compares a hash of the secrets.